libyul/backends/evm/EVMCodeTransform.cpp

```diff
@@ -628,13 +655,72 @@ void CodeTransform::visitExpression(Expression const& _expression)
 	expectDeposit(1, height);
 }
 
+void CodeTransform::setupReturnVariablesAndFunctionExit()
+{
+	yulAssert(!m_functionExitStackHeight.has_value(), "");
+	yulAssert(m_scope, "");
+
+	ScopeGuard scopeGuard([oldScope = m_scope, this] { m_scope = oldScope; });
+	if (!m_scope->functionScope)
+	{
+		yulAssert(m_scope->superScope && m_scope->superScope->functionScope, "");
+		m_scope = m_scope->superScope;
+	}
+
+	if (m_delayedReturnVariables.empty())
+	{
+		m_functionExitStackHeight = m_evm15 ? 0 : 1;
+		return;
+	}
+
+	// Allocate slots for return variables as if they were declared as variables in the virtual function scope.
+	for (TypedName const& var: m_delayedReturnVariables)
+		(*this)(VariableDeclaration{var.location, {var}, {}});
+
+	m_functionExitStackHeight = ranges::max(m_delayedReturnVariables | ranges::views::transform([&](TypedName const& _name) {
+		return variableStackHeight(_name.name);
+	})) + 1;
+	m_delayedReturnVariables.clear();
+}
+
+namespace
+{
+
+bool statementNeedsReturnVariableSetup(Statement const& _statement, vector<TypedName> const& _returnVariables)
+{
+	if (holds_alternative<FunctionDefinition>(_statement))
+		return false;
+	if (
+		holds_alternative<ExpressionStatement>(_statement) ||
+		holds_alternative<Assignment>(_statement)
```

Comment:
Can you explain again why we cannot delay beyond a variable declaration?

Reply:
Or is it just not required for the "stack to memory mover" and would complicate things?

Reply:
Variable declarations will allocate a stack slot, so this would distort the stack layout in all sorts of ways.
Consider:
```
function f(a) -> r
{
  let l1, ..., l16 := g()
  r := 0
  h(l1, ..., l16)
  sstore(a, 42) // unreachable if I allocate after the variable declaration, but reachable if I allocate before.
}
```

In general the goal is *not* to keep the overall stack size as small as possible, but to have the range of currently live variables to be as compact as possible at all times and if I allocate return variables only after variable declarations I freeze slots that might otherwise be popped.

Even worse, I won't be able to reach the return jump label when reshuffling at the end of the function. ``stackLayout`` in that case will be `1, -1, ..., -1, 0` and hit the `stackLayout.size() > 17` error case.